### PR TITLE
Added progress bar for file handler

### PR DESCRIPTION
**Pull request by ngondalia**
changes made:
- Implemented progress bar for when traces are opened.
- Added an output channel for the extension for logging purposes

This will provide the users with a progress bar when the trace file is opened. (Similar to how theia provides a progress bar)

Signed-off-by: Neel Gondalia ngondalia@blackberry.com

### CODE REVIEW

**Comment by ngondalia** [comment]
Screenshot from 2023-04-14 15-35-11

File: vscode-trace-extension/src/utils/traceExtensionLogger.ts
Comment (bhufmann) [inline vscode-trace-extension/src/utils/traceExtensionLogger.ts:9]: class name start with upper case: `TraceExtensionLogger"

File: vscode-trace-extension/src/utils/traceExtensionLogger.ts
Comment (bhufmann) [inline vscode-trace-extension/src/utils/traceExtensionLogger.ts:1]: Please rename `traceExtensionLogger.ts` to `trace-extension-logger.ts` to be consistent with other source code files

Most of the methods are not used. Could you please add some description to the message to know what each method does?

Questions:
- Does vscode make sure that panel is not overflowing?
- Does the panel need to be hidden or disposed at some point?

**Review by bhufmann — CHANGES_REQUESTED** [review]
Thanks for the contribution! Useful addition. I have some comments though.

File: vscode-trace-extension/src/utils/traceExtensionLogger.ts
Comment (ngondalia) [reply vscode-trace-extension/src/utils/traceExtensionLogger.ts]: I have added descriptions for the methods in the Logger.

Does vscode make sure that panel is not overflowing?
-> yes. There is also a method to clear the log channel both from the API and through the UI

 Does the panel need to be hidden or disposed at some point?
-> I have it disposing on extension deactivation now. Reason being my thinking is that this channel can be used by all the modules of the extension, not just the file handler. As for showing/hiding it, the current API only allows hiding the entire outputs panel, and not the individual channels. Hiding individual channels and replacing it with the previously shown channel would be more ideal but its the current limitation of the vscode api. For now, we can just show the output log channel whenever we log something to it and not worry about hiding it. Other extensions that use output channels would do the same.

->The alternatives are always keeping it hidden, and the user can decide when they want to see the trace extension channel by selecting it from the dropdown menu in the outputs panel. The other option is actually using the hide channel api call, which will hide the entire outputs panel (not my preference).

WDYT?

File: vscode-trace-extension/src/trace-explorer/trace-tree.ts
Comment (bhufmann) [inline vscode-trace-extension/src/trace-explorer/trace-tree.ts:124]: This is not an error and we should not log it. traceUri is undefined when the user cancels the dialog.

File: vscode-trace-extension/src/utils/traceExtensionLogger.ts
Comment (bhufmann) [reply vscode-trace-extension/src/utils/traceExtensionLogger.ts]: I think opening the log window everytime we log something might be a bit disruptive, because I would have to close it everytime manually when I want to get more vertical space. I understand that it would be less discoverable for end-users, but I think it's better. I don't think other extension open it automatically.

File: vscode-trace-extension/src/utils/trace-extension-logger.ts
Comment (bhufmann) [inline vscode-trace-extension/src/utils/trace-extension-logger.ts:138]: We should log also the time. I see other plug-ins doing the same. I will be useful if we log more things and want to trouble-shoot.

**Review by bhufmann — CHANGES_REQUESTED** [review]

File: vscode-trace-extension/src/utils/traceExtensionLogger.ts
Comment (ngondalia) [reply vscode-trace-extension/src/utils/traceExtensionLogger.ts]: yeah i agree with that. We can just log to it, and if user wants to inspect it, they can open the channel.

File: vscode-trace-extension/src/utils/trace-extension-logger.ts
Comment (ngondalia) [reply vscode-trace-extension/src/utils/trace-extension-logger.ts:138]: It does have time.. did you mean date?

ss1

Comment (bhufmann) [reply vscode-trace-extension/src/utils/trace-extension-logger.ts:138]: Oh, true, the date is missing not the time. Right now we don't need more details. But you could change it right away for the future.  I checked a few outputs from other extension and most show something like this: 2023-04-20 13:37:24.500.

Comment (ngondalia) [reply vscode-trace-extension/src/utils/trace-extension-logger.ts]: I have modified the timestamp to be in the same format as the one you mentioned

File: vscode-trace-extension/src/utils/trace-extension-logger.ts
Comment (bhufmann) [inline vscode-trace-extension/src/utils/trace-extension-logger.ts:139]: the last 2 lines are not needed

File: vscode-trace-extension/src/utils/trace-extension-logger.ts
Comment (bhufmann) [inline vscode-trace-extension/src/utils/trace-extension-logger.ts:1]: The file uses other format than other source code file, e.g. 4 space for indention. Could you please format the file? Using the vscode / vscodium formatter should work.

**Review by bhufmann — CHANGES_REQUESTED** [review]
Works fine Small nits left to fix.

**Review by bhufmann — APPROVED** [review]
Looks good to me. Thanks for the contribution.

File: vscode-trace-extension/src/extension.ts
Comment (marco-miller) [inline vscode-trace-extension/src/extension.ts:9]: Can we split this PR into at least two commits? Say, one for this logging added, and another for the progress bar addition. If there are refactorings not pertaining to either one of these two main topics, then that should be at least yet another commit.

Doing so enables reviewability, and revertabiliy (should either one of the main changes be deemed problematic after merges). If they all sit in the same commit, then one has to revert everything in case of only some revertworthy bug(s).

Now, if a commit depends on (or builds on top of) another, then it can simply be based on the latter, as usual with git commits indeed.

File: vscode-trace-extension/src/trace-explorer/trace-tree.ts
Comment (marco-miller) [inline vscode-trace-extension/src/trace-explorer/trace-tree.ts:216]: Could this duplicating message string be a reused constant instead? So that we consistently maintain it if need be; we could otherwise forget one. This may apply also to other non-trivial messages in this change.

In fact, maybe rather extract this whole duplicated rollbacking logic in a reused function?

File: vscode-trace-extension/src/trace-explorer/trace-tree.ts
Comment (marco-miller) [inline vscode-trace-extension/src/trace-explorer/trace-tree.ts:185]: I'd remove this one-liner comment, as we already state the matter thanks to line 186's message right below.

File: vscode-trace-extension/src/trace-explorer/trace-tree.ts
Comment (marco-miller) [inline vscode-trace-extension/src/trace-explorer/trace-tree.ts:215]: Same as above, here too.

File: vscode-trace-extension/src/utils/trace-extension-logger.ts
Comment (marco-miller) [inline vscode-trace-extension/src/utils/trace-extension-logger.ts:91]: "Shows `an` information message", rather.

File: vscode-trace-extension/src/utils/trace-extension-logger.ts
Comment (marco-miller) [inline vscode-trace-extension/src/utils/trace-extension-logger.ts:92]: "message to the log channel", rather?

File: vscode-trace-extension/src/utils/trace-extension-logger.ts
Comment (marco-miller) [inline vscode-trace-extension/src/utils/trace-extension-logger.ts:106]: Same as earlier.

File: vscode-trace-extension/src/utils/trace-extension-logger.ts
Comment (marco-miller) [inline vscode-trace-extension/src/utils/trace-extension-logger.ts:120]: Same here too.

File: vscode-trace-extension/src/utils/trace-extension-logger.ts
Comment (marco-miller) [inline vscode-trace-extension/src/utils/trace-extension-logger.ts:134]: The previous function docs have a blank line before their `@param` tags.

Applies to the following ones below and down.

**Review by marco-miller — CHANGES_REQUESTED** [review]
Thanks, yes; some more comments please.

File: vscode-trace-extension/src/extension.ts
Comment (ngondalia) [reply vscode-trace-extension/src/extension.ts:9]: Done! https://github.com/eclipse-cdt-cloud/vscode-trace-extension/pull/129

Once the logging commit is merged, i will update this PR. Thanks!

**Comment by ngondalia** [comment]
Depends on #129

**Comment by marco-miller** [comment]
> Depends on #129

You could also base this commit on #129's, say, in that PR (branch) solely. Thus having two commits in that PR. Then amend the PR description so it combines both speeches, logically speaking. But each commit would remain separate with its focused commit message, indeed. Us reviewers may then review /try each commit, step-wise. This approach has the advantage of preserving the dependency context altogether, while allowing for single commits to exist and be reviewable.

You'd also eventually need to individually Resolve or reply to each open comment above; thanks. This is so that I (the Reviewer here) know when to review again.

**Comment by ngondalia** [comment]
@bhufmann @marco-miller It should be good now. Please review whenever you get a chance! Thanks!

File: vscode-trace-extension/src/trace-explorer/trace-tree.ts
Comment (marco-miller) [inline vscode-trace-extension/src/trace-explorer/trace-tree.ts:226]: `= async`

File: vscode-trace-extension/src/trace-explorer/trace-tree.ts
Comment (marco-miller) [inline vscode-trace-extension/src/trace-explorer/trace-tree.ts:227]: `message:` would suffice, as `undefined` is already specified (both being mutually redundant here).

File: vscode-trace-extension/src/trace-explorer/trace-tree.ts
Comment (marco-miller) [inline vscode-trace-extension/src/trace-explorer/trace-tree.ts:228]: `increment:`, per above.

File: vscode-trace-extension/src/trace-explorer/trace-tree.ts
Comment (marco-miller) [inline vscode-trace-extension/src/trace-explorer/trace-tree.ts:230]: `, increment` (space doubled).

**Review by marco-miller — CHANGES_REQUESTED** [review]
A few minor comments; thanks.

**Review by marco-miller — APPROVED** [review]
LGTM now, thanks. Letting @bhufmann do the second review.

**Review by bhufmann — APPROVED** [review]
Looks good to me! Thanks for the contribution.

**Pull request merged**
